### PR TITLE
feat(testing-framework): bootstrap initial version

### PR DESCRIPTION
**Pull request by jxs**
## Description

Presents the initial version for discussion of the public api.

## Notes
Needs tests, after the public Api discussion is done. 

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by jxs** [comment]
Updated to reflect our discussion of band regarding loading the main actor first and then running multiple contracts, ptal. Meanwhile [I was wrong regarding not being idiomatic the constructor failing](https://users.rust-lang.org/t/should-i-return-a-result-from-new/17372). We can have a constructor `new` accepting the `main_actor` or any other name.

File: testing-framework/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+const EAM_ACTOR_ID: ActorID = 10;
+const NETWORK_VERSION: NetworkVersion = NetworkVersion::V18;
+const DEFAULT_BASE_FEE: u64 = 100;
+
+/// Main interface to test `Actor`s with Kythera.
+pub struct Tester {
+    // Builtin actors root Cid used in the Machine
+    builtin_actors: BuiltInActors,
+    // Custom code cid deployed by developer
+    code_cids: Vec<Cid>,
```
Comment (tchataigner) [inline testing-framework/src/lib.rs:43]: What is the usage of this property ? It's only written right now. I can forsee maybe two of them:
1. External usage, let our user know what are the CID of its deployed actors, both test and target
2. Internal usage, ability to know what are the links between test and target actor

In both cases, I think that differentiating the target and test actors seems like a good idea.

Also, the second case I've written (which is not currently covered I think) seems interesting to have.

File: testing-framework/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+            Ok(id) => Some(id.to_ne_bytes().to_vec()),
+            Err(_) => panic!("Actor Id should be valid"),
+        };
+
+        Ok(())
+    }
+
+    /// Test an Actor on a `MemoryBlockstore`.
+    pub fn test(&mut self, actor: WasmActor, test: WasmActor) -> Result<(), Error> {
+        // TODO: Should we clone the `StateTree` before each test run,
```
Comment (tchataigner) [inline testing-framework/src/lib.rs:302]: Having both an `actor` and a `test` parameters here seems redundant, at the very least for the use case we make of them right now (only logging). I think we should:
1. Store names in our tester. 
2. Use the actor ID of the `actor` to refer to it and get its data accordingly. 

_Note: The unique IDs should most likely be actors' ids_

File: testing-framework/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+/// Built-in Actors that are deployed to the testing `StateTree`.
+struct BuiltInActors {
+    root: Cid,
+    manifest: Manifest,
+}
+
+/// WebAssembly Actor.
+pub struct WasmActor {
+    name: String,
+    code: Vec<u8>,
```
Comment (tchataigner) [inline testing-framework/src/lib.rs:61]: Wouldn't some code be most likely be `&[u8]` instead ?

File: testing-framework/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    fn deploy_actor_from_bin(
+        &mut self,
+        actor: &WasmActor,
+        balance: TokenAmount,
+    ) -> Result<Address, Error> {
+        let actor_id = rand::random();
+        let actor_address = Address::new_id(actor_id);
+
+        // Put the WASM code into the blockstore.
+        log::debug!("Deploying Actor {} code ", actor.name);
```
Comment (tchataigner) [inline testing-framework/src/lib.rs:259]:
```suggestion
        log::debug!("Deploying Actor {} code", actor.name);
```

File: testing-framework/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+            code_cids: vec![],
+            state_tree,
+            account,
+            main_actor_id: None,
+        }
+    }
+
+    /// Load the built-in actors into the `Blockstore`.
+    /// And activate them on the `StateTree`.
+    fn load_builtin_actors<B: Blockstore>(state_tree: &mut StateTree<B>) -> BuiltInActors {
```
Comment (tchataigner) [inline testing-framework/src/lib.rs:96]: I think this function is incomplete. We are missing quite a few actors from the builtins that need to be instantiated.

We could refer to [Workbench](https://github.com/anorth/fvm-workbench/blob/main/builtin/src/genesis.rs) to help us on that.

File: testing-framework/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+use error::{Error, WrapFVMError};
+use fvm_shared::{
+    address::Address, bigint::Zero, econ::TokenAmount, message::Message, state::StateTreeVersion,
+    version::NetworkVersion, ActorID, IPLD_RAW,
+};
+
+// TODO: document purpose.
+const EAM_ACTOR_ID: ActorID = 10;
+const NETWORK_VERSION: NetworkVersion = NetworkVersion::V18;
```
Comment (tchataigner) [inline testing-framework/src/lib.rs:35]: This should be made dynamic I think

File: testing-framework/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+            method_num: 1,
+            params: main_actor_id.into(),
+            ..Message::default()
+        };
+
+        log::info!("testing test {} to Actor {}", test.name, actor.name);
+        executor
+            .execute_message(message, ApplyKind::Explicit, 100)
+            .tester_err(&format!("Could not test the Actor: {}", actor.name))
+            .map(|_| ())
```
Comment (tchataigner) [inline testing-framework/src/lib.rs:335]: For each test actos we will be using against our target actor there will be multiple tests. 

Which means that when we execute the `test()` method we should have a loop somewhere that's calling for all available methods exposed by our test file. 

If what kept you from doing it is the lack of information of which are the methods we should at least have a TODO in the code to tackle it when the time is right.

**Review by tchataigner — CHANGES_REQUESTED** [review]
Hey @jxs ! Thanks for the work, I think before moving on we have 2 main subject to align ourselves on: flow of the test framework and what to do with incomplete part of the code that currently exist.

## Flow

If I understand correctly the flowyou went for is:
1. Instantiate `Tester` (set builtins and generate some acccounts) with `new()`
2. Deploy target actor with `deploy_main_actor()`
3. Call `test()` for each test file against the target actor.

Am I right ? If so, this looks ok to me for now if we tackle a few comments I posted.

## Incomplete work

I think new issues should be created to handle all points where we think we could improve the code as to separate their completion. Would you agree ?

File: testing-framework/src/lib.rs (hunk 2 above)
Comment (jxs) [reply testing-framework/src/lib.rs]: yeah agreed, this was from when we thought that actor and test relationship would be 1:1, updated to match our new 1:N usecase ptal Thomas.

File: testing-framework/src/lib.rs (hunk 7 above)
Comment (jxs) [reply testing-framework/src/lib.rs]: yup, in line with the above comment, updated to reflect the updated usecase

File: testing-framework/src/lib.rs (hunk 3 above)
Comment (jxs) [reply testing-framework/src/lib.rs]: have you started on this Thomas (after our conversation)? If not I can take it on.

File: testing-framework/src/lib.rs (hunk 1 above)
Comment (jxs) [reply testing-framework/src/lib.rs]: glad you mention! I was going to ask it, I removed it meanwhile, tbh I'd suggest we remove it for now as we don't have a use yet (YAGNI) and add it when we need, wdyt?

File: testing-framework/src/lib.rs (hunk 6 above)
Comment (jxs) [reply testing-framework/src/lib.rs]: updated!

File: Cargo.toml [hunk 8]
```diff
@@ -2,4 +2,5 @@
 
 members = [
     "cli",
+    "testing-framework",
```
[the remark this answers is not in the document]
Comment (jxs) [reply Cargo.toml]: discussed off band, and Thomas addressed it in this PR

**Review by tchataigner — DISMISSED** [review]
LGTM !

**Comment by tchataigner** [comment]
@jxs tests seem to fail in CI

**Review by tchataigner — APPROVED** [review]
Should be alright to merge now

**Pull request merged**